### ISSUE TITLE
Race condition in test

### ISSUE THREAD

**Issue by paulbellamy** [weaveworks/scope#648]
Could indicate a real race condition.

```
--- FAIL: TestControl (0.12s)
panic: sync: WaitGroup is reused before previous Wait has returned [recovered]
	panic: sync: WaitGroup is reused before previous Wait has returned

goroutine 84 [running]:
testing.tRunner.func1(0xc8202b2d80)
	/usr/local/go/src/testing/testing.go:450 +0x171
sync.(*WaitGroup).Wait(0xc820492d30)
	/usr/local/go/src/sync/waitgroup.go:128 +0x114
net/http/httptest.(*Server).Close(0xc820492d00)
	/usr/local/go/src/net/http/httptest/server.go:168 +0x5c
github.com/weaveworks/scope/app.TestControl(0xc8202b2d80)
	/home/vagrant/src/github.com/weaveworks/scope/app/controls_test.go:69 +0xadb
testing.tRunner(0xc8202b2d80, 0x1c1b0b8)
	/usr/local/go/src/testing/testing.go:456 +0x98
created by testing.RunTests
	/usr/local/go/src/testing/testing.go:561 +0x86d

goroutine 1 [chan receive]:
testing.RunTests(0x14183b8, 0x1c1afe0, 0xc, 0xc, 0xc820292b00)
	/usr/local/go/src/testing/testing.go:562 +0x8ad
testing.(*M).Run(0xc8202cdef8, 0x7fcc44085038)
	/usr/local/go/src/testing/testing.go:494 +0x70
main.main()
	github.com/weaveworks/scope/app/_test/_testmain.go:76 +0x116

goroutine 17 [syscall, locked to thread]:
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1696 +0x1

goroutine 20 [chan receive]:
github.com/weaveworks/scope/vendor/github.com/golang/glog.(*loggingT).flushDaemon(0x1c39ae0)
	/home/vagrant/src/github.com/weaveworks/scope/vendor/github.com/golang/glog/glog.go:882 +0x67
created by github.com/weaveworks/scope/vendor/github.com/golang/glog.init.1
	/home/vagrant/src/github.com/weaveworks/scope/vendor/github.com/golang/glog/glog.go:410 +0x297

goroutine 22 [syscall]:
os/signal.loop()
	/usr/local/go/src/os/signal/signal_unix.go:22 +0x18
created by os/signal.init.1
	/usr/local/go/src/os/signal/signal_unix.go:28 +0x37

goroutine 63 [IO wait]:
net.runtime_pollWait(0x7fcc444e8048, 0x72, 0xc820068080)
	/usr/local/go/src/runtime/netpoll.go:157 +0x60
net.(*pollDesc).Wait(0xc8201220d0, 0x72, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:73 +0x3a
net.(*pollDesc).WaitRead(0xc8201220d0, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:78 +0x36
net.(*netFD).Read(0xc820122070, 0xc820414000, 0x1000, 0x1000, 0x0, 0x7fcc444e2028, 0xc820068080)
	/usr/local/go/src/net/fd_unix.go:232 +0x23a
net.(*conn).Read(0xc820498088, 0xc820414000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/net.go:172 +0xe4
net/http.noteEOFReader.Read(0x7fcc44085130, 0xc820498088, 0xc820412058, 0xc820414000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/http/transport.go:1370 +0x67
net/http.(*noteEOFReader).Read(0xc82058e760, 0xc820414000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	<autogenerated>:126 +0xd0
bufio.(*Reader).fill(0xc820408300)
	/usr/local/go/src/bufio/bufio.go:97 +0x1e9
bufio.(*Reader).Peek(0xc820408300, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0)
	/usr/local/go/src/bufio/bufio.go:132 +0xcc
net/http.(*persistConn).readLoop(0xc820412000)
	/usr/local/go/src/net/http/transport.go:876 +0xf7
created by net/http.(*Transport).dialConn
	/usr/local/go/src/net/http/transport.go:685 +0xc78

goroutine 64 [select]:
net/http.(*persistConn).writeLoop(0xc820412000)
	/usr/local/go/src/net/http/transport.go:1009 +0x40c
created by net/http.(*Transport).dialConn
	/usr/local/go/src/net/http/transport.go:686 +0xc9d

goroutine 72 [IO wait]:
net.runtime_pollWait(0x7fcc444e8108, 0x72, 0xc820068080)
	/usr/local/go/src/runtime/netpoll.go:157 +0x60
net.(*pollDesc).Wait(0xc82025e680, 0x72, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:73 +0x3a
net.(*pollDesc).WaitRead(0xc82025e680, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:78 +0x36
net.(*netFD).accept(0xc82025e620, 0x0, 0x7fcc440850e0, 0xc8204a28e0)
	/usr/local/go/src/net/fd_unix.go:408 +0x27c
net.(*TCPListener).AcceptTCP(0xc82006e068, 0xc820492b00, 0x0, 0x0)
	/usr/local/go/src/net/tcpsock_posix.go:254 +0x4d
net.(*TCPListener).Accept(0xc82006e068, 0x0, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/tcpsock_posix.go:264 +0x3d
net/http/httptest.(*historyListener).Accept(0xc82027c750, 0x0, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/http/httptest/server.go:48 +0x63
net/http.(*Server).Serve(0xc8202f4780, 0x7fcc444e8348, 0xc82027c750, 0x0, 0x0)
	/usr/local/go/src/net/http/server.go:1887 +0xb3
created by net/http/httptest.(*Server).Start
	/usr/local/go/src/net/http/httptest/server.go:109 +0x380

goroutine 65 [IO wait]:
net.runtime_pollWait(0x7fcc444e7f88, 0x72, 0xc820068080)
	/usr/local/go/src/runtime/netpoll.go:157 +0x60
net.(*pollDesc).Wait(0xc820122220, 0x72, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:73 +0x3a
net.(*pollDesc).WaitRead(0xc820122220, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:78 +0x36
net.(*netFD).Read(0xc8201221c0, 0xc820212000, 0x1000, 0x1000, 0x0, 0x7fcc444e2028, 0xc820068080)
	/usr/local/go/src/net/fd_unix.go:232 +0x23a
net.(*conn).Read(0xc8204980b8, 0xc820212000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/net.go:172 +0xe4
net/http.(*liveSwitchReader).Read(0xc8204120f8, 0xc820212000, 0x1000, 0x1000, 0xc820125000, 0x0, 0x0)
	/usr/local/go/src/net/http/server.go:219 +0xa4
io.(*LimitedReader).Read(0xc82058e820, 0xc820212000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/local/go/src/io/io.go:427 +0xbd
bufio.(*Reader).fill(0xc8202abaa0)
	/usr/local/go/src/bufio/bufio.go:97 +0x1e9
bufio.(*Reader).Peek(0xc8202abaa0, 0x4, 0x0, 0x0, 0x0, 0x0, 0x0)
	/usr/local/go/src/bufio/bufio.go:132 +0xcc
net/http.(*conn).readRequest(0xc8204120b0, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/http/server.go:629 +0x2bb
net/http.(*conn).serve(0xc8204120b0)
	/usr/local/go/src/net/http/server.go:1319 +0x727
created by net/http.(*Server).Serve
	/usr/local/go/src/net/http/server.go:1910 +0x3f6

goroutine 44 [chan receive]:
github.com/weaveworks/scope/app.(*controlRouter).handleProbeWS(0xc820285bb0, 0x7fcc444e8560, 0xc820106420, 0xc820432000)
	/home/vagrant/src/github.com/weaveworks/scope/app/controls.go:117 +0x3c1
github.com/weaveworks/scope/app.(*controlRouter).(github.com/weaveworks/scope/app.handleProbeWS)-fm(0x7fcc444e8560, 0xc820106420, 0xc820432000)
	/home/vagrant/src/github.com/weaveworks/scope/app/controls.go:19 +0x3e
net/http.HandlerFunc.ServeHTTP(0xc820285fe0, 0x7fcc444e8560, 0xc820106420, 0xc820432000)
	/usr/local/go/src/net/http/server.go:1422 +0x3a
github.com/weaveworks/scope/vendor/github.com/gorilla/mux.(*Router).ServeHTTP(0xc820476a50, 0x7fcc444e8560, 0xc820106420, 0xc820432000)
	/home/vagrant/src/github.com/weaveworks/scope/vendor/github.com/gorilla/mux/mux.go:100 +0x29e
net/http/httptest.(*waitGroupHandler).ServeHTTP(0xc82012f200, 0x7fcc444e8560, 0xc820106420, 0xc820432000)
	/usr/local/go/src/net/http/httptest/server.go:200 +0xc2
net/http.serverHandler.ServeHTTP(0xc8202f5e00, 0x7fcc444e8560, 0xc820106420, 0xc820432000)
	/usr/local/go/src/net/http/server.go:1862 +0x19e
net/http.(*conn).serve(0xc820106370)
	/usr/local/go/src/net/http/server.go:1361 +0xbee
created by net/http.(*Server).Serve
	/usr/local/go/src/net/http/server.go:1910 +0x3f6

goroutine 87 [IO wait]:
net.runtime_pollWait(0x7fcc444e7ec8, 0x72, 0xc820068080)
	/usr/local/go/src/runtime/netpoll.go:157 +0x60
net.(*pollDesc).Wait(0xc82048c610, 0x72, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:73 +0x3a
net.(*pollDesc).WaitRead(0xc82048c610, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:78 +0x36
net.(*netFD).Read(0xc82048c5b0, 0xc820438000, 0x1000, 0x1000, 0x0, 0x7fcc444e2028, 0xc820068080)
	/usr/local/go/src/net/fd_unix.go:232 +0x23a
net.(*conn).Read(0xc8200cc0e0, 0xc820438000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/net.go:172 +0xe4
bufio.(*Reader).fill(0xc820064ba0)
	/usr/local/go/src/bufio/bufio.go:97 +0x1e9
bufio.(*Reader).Read(0xc820064ba0, 0xc8203e8be0, 0x2, 0x8, 0xc8203de960, 0x0, 0x0)
	/usr/local/go/src/bufio/bufio.go:207 +0x260
github.com/weaveworks/scope/vendor/github.com/gorilla/websocket.(*Conn).readFull(0xc82043a000, 0xc8203e8be0, 0x2, 0x8, 0x0, 0x0)
	/home/vagrant/src/github.com/weaveworks/scope/vendor/github.com/gorilla/websocket/conn.go:546 +0xa9
github.com/weaveworks/scope/vendor/github.com/gorilla/websocket.(*Conn).advanceFrame(0xc82043a000, 0xc81ffe1095, 0x0, 0x0)
	/home/vagrant/src/github.com/weaveworks/scope/vendor/github.com/gorilla/websocket/conn.go:570 +0x159
github.com/weaveworks/scope/vendor/github.com/gorilla/websocket.(*Conn).NextReader(0xc82043a000, 0xc8203ded50, 0x0, 0x0, 0x0, 0x0)
	/home/vagrant/src/github.com/weaveworks/scope/vendor/github.com/gorilla/websocket/conn.go:705 +0x76
github.com/weaveworks/scope/vendor/github.com/gorilla/websocket.(*Conn).ReadJSON(0xc82043a000, 0xe00fc0, 0xc8203ded50, 0x0, 0x0)
	/home/vagrant/src/github.com/weaveworks/scope/vendor/github.com/gorilla/websocket/json.go:45 +0x37
github.com/weaveworks/scope/xfer.(*JSONWebsocketCodec).ReadResponseHeader(0xc8203e5240, 0xc8203ded50, 0x0, 0x0)
	/home/vagrant/src/github.com/weaveworks/scope/xfer/controls.go:93 +0x55
net/rpc.(*Client).input(0xc8203c0960)
	/usr/local/go/src/net/rpc/client.go:109 +0xbf
created by net/rpc.NewClientWithCodec
	/usr/local/go/src/net/rpc/client.go:201 +0xd2

goroutine 86 [runnable]:
syscall.Syscall(0x0, 0x6, 0xc8203f1000, 0x1000, 0x81, 0x1000, 0x0)
	/usr/local/go/src/syscall/asm_linux_amd64.s:18 +0x5
syscall.read(0x6, 0xc8203f1000, 0x1000, 0x1000, 0x72, 0x0, 0x0)
	/usr/local/go/src/syscall/zsyscall_linux_amd64.go:783 +0x5f
syscall.Read(0x6, 0xc8203f1000, 0x1000, 0x1000, 0xffffffffffffff01, 0x0, 0x0)
	/usr/local/go/src/syscall/syscall_unix.go:160 +0x4d
net.(*netFD).Read(0xc820157730, 0xc8203f1000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/fd_unix.go:228 +0x18b
net.(*conn).Read(0xc82006e0f0, 0xc8203f1000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/net.go:172 +0xe4
bufio.(*Reader).fill(0xc8203c08a0)
	/usr/local/go/src/bufio/bufio.go:97 +0x1e9
bufio.(*Reader).ReadSlice(0xc8203c08a0, 0xc81ffd4a0a, 0x0, 0x0, 0x0, 0x0, 0x0)
	/usr/local/go/src/bufio/bufio.go:328 +0x21a
bufio.(*Reader).ReadLine(0xc8203c08a0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/usr/local/go/src/bufio/bufio.go:357 +0x53
net/textproto.(*Reader).readLineSlice(0xc8200c1820, 0x0, 0x0, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/textproto/reader.go:55 +0x81
net/textproto.(*Reader).ReadLine(0xc8200c1820, 0x0, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/textproto/reader.go:36 +0x40
net/http.ReadResponse(0xc8203c08a0, 0xc8203b3180, 0xc82006e0f0, 0x0, 0x0)
	/usr/local/go/src/net/http/response.go:128 +0xf6
github.com/weaveworks/scope/vendor/github.com/gorilla/websocket.(*Dialer).Dial(0xc8203e0f50, 0xc8203dec30, 0x23, 0xc8203debd0, 0x0, 0x0, 0x0, 0x0)
	/home/vagrant/src/github.com/weaveworks/scope/vendor/github.com/gorilla/websocket/client.go:304 +0x156b
github.com/weaveworks/scope/xfer.(*appClient).controlConnection(0xc82056af30, 0x7fcc45b34370, 0xc8203e8840, 0x0, 0x0)
	/home/vagrant/src/github.com/weaveworks/scope/xfer/app_client.go:90 +0x158
github.com/weaveworks/scope/xfer.(*appClient).controlConnectionLoop(0xc82056af30, 0x7fcc45b34370, 0xc8203e8840)
	/home/vagrant/src/github.com/weaveworks/scope/xfer/app_client.go:130 +0x168
created by github.com/weaveworks/scope/xfer.(*appClient).ControlConnection
	/home/vagrant/src/github.com/weaveworks/scope/xfer/app_client.go:150 +0x49
FAIL	github.com/weaveworks/scope/app	0.484s
```

**Comment by paulbellamy** [comment]
Based on reading `net/http/httptest/server.go`, and `sync/waitgroup_test.go` sources, it seems that a likely cause is that there is a request being made to the test server while it is being shut down.

**Comment by tomwilkie** [comment]
I've got to the bottom of this, and it will me fixed when #650 is merged.
